nexus/pkg/server/stream.go

```diff
@@ -117,33 +125,43 @@ func (s *Stream) GetRun() *service.RunRecord {
 	return s.handler.GetRun()
 }
 
-// Close closes the stream's handler, writer, sender, and dispatcher.
-// This can be triggered by the client (force=False) or by the server (force=True).
-// We need the ExitRecord to initiate the shutdown procedure (which we
-// either get from the client, or generate ourselves if the server is shutting us down).
-// This will trigger the defer state machines (SM) in the stream's components:
-//   - when the sender's SM gets to the final state, it will Close the handler
-//   - this will trigger the handler to Close the writer
-//   - this will trigger the writer to Close the sender
-//
-// This will finish the Stream's wait group, which will allow the stream to be
-// garbage collected.
+// Gracefully wait for handler, writer, sender, dispatcher to shutdown cleanly
+// assumes an exit record has already been sent
 func (s *Stream) Close() {
-	// Close and wait for input channel to shutdown
-	close(s.inChan)
 	s.wg.Wait()
 }
 
+// Handle internal responses like from the finish and close path
+func (s *Stream) Respond(resp *service.ServerResponse) {
+	s.respChan <- resp
+}
+
 func (s *Stream) FinishAndClose(exitCode int32) {
+	s.AddResponders(ResponderEntry{s, "internal"})
+
 	// send exit record to handler
 	record := &service.Record{
 		RecordType: &service.Record_Exit{
 			Exit: &service.RunExitRecord{
 				ExitCode: exitCode,
 			}},
-		Control: &service.Control{AlwaysSend: true},
+		Control: &service.Control{AlwaysSend: true, ConnectionId: "internal", ReqResp: true},
 	}
+
 	s.HandleRecord(record)
+	<-s.respChan
+
+	shutdownRecord := &service.Record{
+		RecordType: &service.Record_Request{
+			Request: &service.Request{
+				RequestType: &service.Request_Shutdown{
+					Shutdown: &service.ShutdownRequest{},
+				},
+			}},
+		Control: &service.Control{AlwaysSend: true, ConnectionId: "internal", ReqResp: true},
```

Comment:
see above

Reply:
yep, definitely needed... not super clean, but needed

The teardown path, we are issuing a series of requests (right now just 2) as if we are another client of the stream.   We are pushing our requests into the stream and we need the responses, hence why we add a responder at the beginning of the function